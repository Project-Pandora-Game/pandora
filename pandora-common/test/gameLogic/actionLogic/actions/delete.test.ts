import { describe, expect, it } from '@jest/globals';
import { AssertNotNullable } from '../../../../src/index.ts';
import { TestAssetsLoadAssetManager } from '../../../assets/_testData/testAssetsDefinition.ts';
import { TestCreateCharacterState, TestCreateGlobalState } from '../../../assets/assetsTestingHelpers.ts';
import { TestCreateGameLogicCharacter } from '../../character/characterTestingHelpers.ts';
import { TestActionExpectFail, TestActionExpectValidResult, TestCreateActionContext, TestDoImmediateAction, TestStateExtractAssets } from '../actionTestingHelpers.ts';

describe('ActionDelete', () => {
	it('Deletes worn item', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, [
			TestCreateCharacterState(assetManager, character, [
				'a/panties/style1',
				'a/panties/style1',
				'a/headwear/top_hat',
			]),
		]);

		const secondPantiesId = baseState.characters.get(character.id)?.items.findLast((i) => i.asset.id === 'a/headwear/top_hat')?.id;
		AssertNotNullable(secondPantiesId);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'delete',
				target: {
					type: 'character',
					characterId: character.id,
				},
				item: {
					container: [],
					itemId: secondPantiesId,
				},
			},
			TestCreateActionContext(character),
			baseState,
		);

		// Check the result
		TestActionExpectValidResult(result);
		expect(TestStateExtractAssets(result.resultState.characters.get(character.id)?.items)).toMatchSnapshot();
	});

	it('Deletes optional bodypart item', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, [
			TestCreateCharacterState(assetManager, character, [
				'a/body/front_hair1',
				'a/headwear/top_hat',
			]),
		]);

		const hairId = baseState.characters.get(character.id)?.items.findLast((i) => i.asset.id === 'a/body/front_hair1')?.id;
		AssertNotNullable(hairId);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'delete',
				target: {
					type: 'character',
					characterId: character.id,
				},
				item: {
					container: [],
					itemId: hairId,
				},
			},
			TestCreateActionContext(character),
			baseState,
		);

		// Check the result
		TestActionExpectValidResult(result);
		expect(TestStateExtractAssets(result.resultState.characters.get(character.id)?.items)).toMatchSnapshot();
	});

	it('Fails if manipulating bodypart in space that does not allow it', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, [
			TestCreateCharacterState(assetManager, character, [
				'a/headwear/top_hat',
			]),
		]);

		const eyesId = baseState.characters.get(character.id)?.items.find((i) => i.asset.id === 'a/body/eyes')?.id;
		AssertNotNullable(eyesId);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'delete',
				target: {
					type: 'character',
					characterId: character.id,
				},
				item: {
					container: [],
					itemId: eyesId,
				},
			},
			TestCreateActionContext(character),
			baseState,
		);

		// Check the result
		TestActionExpectFail(result);
	});

	it('Fails if deleting required bodypart', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, [
			TestCreateCharacterState(assetManager, character, [
				'a/headwear/top_hat',
			]),
		]);

		const eyesId = baseState.characters.get(character.id)?.items.find((i) => i.asset.id === 'a/body/eyes')?.id;
		AssertNotNullable(eyesId);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'delete',
				target: {
					type: 'character',
					characterId: character.id,
				},
				item: {
					container: [],
					itemId: eyesId,
				},
			},
			TestCreateActionContext(character, { spaceFeatures: ['allowBodyChanges'] }),
			baseState,
		);

		// Check the result
		TestActionExpectFail(result);
	});

	it('Fails if deleting nonexistent item', () => {
		const assetManager = TestAssetsLoadAssetManager();
		const character = TestCreateGameLogicCharacter(1, 'c1');
		const baseState = TestCreateGlobalState(assetManager, [
			TestCreateCharacterState(assetManager, character),
		]);

		// Do the action
		const result = TestDoImmediateAction(
			{
				type: 'delete',
				target: {
					type: 'character',
					characterId: character.id,
				},
				item: {
					container: [],
					itemId: 'i/i-definitely-do-not-exist',
				},
			},
			TestCreateActionContext(character),
			baseState,
		);

		// Check the result
		TestActionExpectFail(result);
	});
});
