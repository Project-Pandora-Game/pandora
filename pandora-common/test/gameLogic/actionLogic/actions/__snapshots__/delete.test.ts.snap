// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionDelete Deletes optional bodypart item 1`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/headwear/top_hat",
  },
]
`;

exports[`ActionDelete Deletes worn item 1`] = `
[
  {
    "asset": "a/body/base",
  },
  {
    "asset": "a/body/head",
  },
  {
    "asset": "a/body/eyes",
  },
  {
    "asset": "a/body/lips",
  },
  {
    "asset": "a/panties/style1",
  },
  {
    "asset": "a/panties/style1",
  },
]
`;

exports[`ActionDelete Fails if deleting nonexistent item: problems 1`] = `
[
  {
    "restriction": {
      "type": "invalid",
    },
    "result": "restrictionError",
  },
  {
    "reason": undefined,
    "result": "invalidAction",
  },
]
`;

exports[`ActionDelete Fails if deleting required bodypart: problems 1`] = `
[
  {
    "result": "validationError",
    "validationError": {
      "problem": "bodypartError",
      "problemDetail": "missingRequired",
    },
  },
]
`;

exports[`ActionDelete Fails if manipulating bodypart in space that does not allow it: problems 1`] = `
[
  {
    "restriction": {
      "type": "modifyBodyRoom",
    },
    "result": "restrictionError",
  },
  {
    "result": "validationError",
    "validationError": {
      "problem": "bodypartError",
      "problemDetail": "missingRequired",
    },
  },
]
`;
